pkg/agent/controller/servicecidr/provider_test.go

```diff
@@ -0,0 +1,180 @@
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes/fake"
+)
+
+var (
+	testServiceIPv4CIDR *net.IPNet
+	testServiceIPv6CIDR *net.IPNet
+
+	serviceIPv4CIDRHandlerCalled = 0
+	serviceIPv6CIDRHandlerCalled = 0
+
+	serviceCIDRProvider Interface
+)
+
+func makeService(namespace, name string, clusterIP string, protocol corev1.Protocol) *corev1.Service {
+	svc := &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        name,
+			Namespace:   namespace,
+			Annotations: map[string]string{},
+		},
+		Spec: corev1.ServiceSpec{
+			ClusterIPs: []string{clusterIP},
+			ClusterIP:  clusterIP,
+			Ports: []corev1.ServicePort{
+				{
+					Name:     "80",
+					Port:     int32(80),
+					Protocol: protocol,
+				},
+			},
+			Type: corev1.ServiceTypeClusterIP,
+		},
+	}
+	return svc
+}
+
+func serviceIPv4CIDRHandler() {
+	testServiceIPv4CIDR, _ = serviceCIDRProvider.GetServiceCIDR(false)
+	serviceIPv4CIDRHandlerCalled++
+}
+
+func serviceIPv6CIDRHandler() {
+	testServiceIPv6CIDR, _ = serviceCIDRProvider.GetServiceCIDR(true)
+	serviceIPv6CIDRHandlerCalled++
+}
+
+func TestServiceCIDRProvider(t *testing.T) {
+	client := fake.NewSimpleClientset()
+	informerFactory := informers.NewSharedInformerFactory(client, 0)
+	svcInformer := informerFactory.Core().V1().Services()
+	serviceCIDRProvider = NewServiceCIDRProvider(svcInformer)
+	serviceCIDRProvider.AddEventHandler(serviceIPv4CIDRHandler)
+	serviceCIDRProvider.AddEventHandler(serviceIPv6CIDRHandler)
+
+	stopCh := make(chan struct{})
+	defer close(stopCh)
+	informerFactory.Start(stopCh)
+	informerFactory.WaitForCacheSync(stopCh)
+
+	svc := makeService("ns1", "svc-ipv4-1", "10.10.0.1", corev1.ProtocolTCP)
+	_, err := client.CoreV1().Services("ns1").Create(context.TODO(), svc, metav1.CreateOptions{})
+	assert.NoError(t, err)
+	assert.NoError(t, checkServiceCIDR("10.10.0.1/32", false))
+	// Check the value modified by handler serviceIPv4CIDRHandler.
+	assert.Equal(t, "10.10.0.1/32", testServiceIPv4CIDR.String())
+	assert.Equal(t, 1, serviceIPv4CIDRHandlerCalled)
+
+	svc = makeService("ns1", "svc-ipv4-2", "10.10.0.2", corev1.ProtocolTCP)
+	_, err = client.CoreV1().Services("ns1").Create(context.TODO(), svc, metav1.CreateOptions{})
+	assert.NoError(t, err)
+	assert.NoError(t, checkServiceCIDR("10.10.0.0/30", false))
+	// Check the value modified by handler serviceIPv4CIDRHandler.
+	assert.Equal(t, "10.10.0.0/30", testServiceIPv4CIDR.String())
+	assert.Equal(t, 2, serviceIPv4CIDRHandlerCalled)
+
+	svc = makeService("ns1", "svc-ipv4-5", "10.10.0.5", corev1.ProtocolTCP)
+	_, err = client.CoreV1().Services("ns1").Create(context.TODO(), svc, metav1.CreateOptions{})
+	assert.NoError(t, err)
+	assert.NoError(t, checkServiceCIDR("10.10.0.0/29", false))
+	// Handler serviceIPv4CIDRHandler should not be called.
+	assert.Equal(t, "10.10.0.0/29", testServiceIPv4CIDR.String())
+	assert.Equal(t, 3, serviceIPv4CIDRHandlerCalled)
+
+	svc = makeService("ns1", "svc-ipv4-4", "10.10.0.4", corev1.ProtocolTCP)
```

Comment:
Any reason to use `"svc-ipv4-4"` after `"svc-ipv4-5"` creation?  if there is no special meaning, better to just call them with sequential order.

Reply:
Consistent with IP.